### ISSUE TITLE
Provide API to the editor to decide which form field/component types are exposed in form creation

### ISSUE THREAD

**Issue by christian-konrad** [bpmn-io/form-js#237]
__Is your feature request related to a problem? Please describe__

To allow extensibility in the core of form.js, the embedding client needs control over which form fields can be used when editing and creating forms. This allows a stable set of exposed form field types even when new field types are added to form.js.

This enables innovating on form.js by adding new components without the risk of exposing unwanted components to already embedding clients on a version bump.

__Describe the solution you'd like__

On form editor instantiation, the developer should be able to provide a whitelist of form field/component types that should be exposed in the editor.

Something like
```js
const formEditor = new FormEditor({
  container: document.querySelector('#form-editor'),
  availableComponentTypes: ['textfield', 'number',  'checkbox', ...]
});
```

**Comment by nikku** [comment]
Is this a duplicate of https://github.com/bpmn-io/form-js/issues/123?

**Comment by dfulgham** [comment]
Got this working so far, but what do you want it to do if the imported Schema contains not valid components?

```
const formEditor = formEditorRef.current = new FormEditor({
      renderer: {
        compact: true
      },
      availableComponentTypes:['button','select']
    });
```

in the example below I only passed in 'button' and 'select'

![image](https://user-images.githubusercontent.com/1512108/158713478-d169d441-68bc-40de-9591-0b57e708d598.png)


Also created types:

```

/** from types.d.ts */

export type ComponentTypes =
  'checkbox'|
  'number'|
  'radio'|
  'select'|
  'textfield'|
  'text' |
  'button'

/** from FormEditor,js */

@typedef { {
 *   properties: FormEditorProperties,
 *   availableComponentTypes: ComponentTypes[]
 *   schema: Schema
 * } } State

/** from Pallete.js */

/**
 * @typedef { import('../../../types').ComponentTypes} ComponentTypes
 *
 * @typedef { {label:string,type:ComponentTypes}} PalletComponents
 */

/**
 * @type { PalletComponents[] } types
 */
const types = [
  {
    label: 'Text Field',
    type: 'textfield'
  },
  {
    label: 'Number',
    type: 'number'
....
]

```

**Comment by christian-konrad** [comment]
> #123

This issue essentially proposes a component white-list as a way to describe a stable API for form.js users/embedders (like Camunda Modeler). This will be an enabler to add an extension API and for extensions themselves. It does not describe the feature request to add extensions.

**Comment by nikku** [comment]
I do see this issue as a simple way to filter what extensions to support. I don't think it is needed short or mid-term.

What I do think is needed is a general way to extend form-js, to supply custom elements (editing) but also serialize them into the schema and support them during viewing.

We'll take your feature request for "file upload" (https://github.com/bpmn-io/form-js/issues/222) and the related PR (https://github.com/bpmn-io/form-js/pull/231) as a reference. If we do our homework as we create an extension mechanism adding such component shall be easily possible.

However we'll not be able to follow up on this in the next weeks. As @christian-konrad mentioned we'll put this on our potential investment topics in the mid-term.

If you want to work on your end to sketch how an extension mechanism for form-js could look like: We're happy to start the discussion and review your contribution. We won't be able to jump on this topic ourselves for the time being.

**Comment by dfulgham** [comment]
> some other options that would be nice, is perhaps adding to the API the ability to decide on how to handle components that are not in the allowed list, i.e. `unavailableComponentAction: 'readonly'|'remove'|'allow' ` etc. Read only would allow changing the position in the form but not the component properties, remove would strip the component from schema, and allow would allow editing of the component but obviously not the ability to add it back if removed.

Readonly action:
<img width="1245" alt="image" src="https://user-images.githubusercontent.com/1512108/158936791-942d7ddd-ce11-456a-b264-985753ee654d.png">

Allow Action:
<img width="1245" alt="image" src="https://user-images.githubusercontent.com/1512108/158936844-4959872e-fefc-44a8-b4ab-fd9d54eef60c.png">

Remove Action (Default, and Camunda Action)
<img width="1245" alt="image" src="https://user-images.githubusercontent.com/1512108/158936951-6de6e07e-22ff-46e8-aefb-15025978849c.png">

**Comment by nikku** [comment]
> Well I'm definitely eager to assist with this project.

Thanks. I'm not sure if we're 100% aligned. 

What I'd start with is #123, an extension mechanism that allows a field such as file upload to be provided.

As a follow up, I'd look into this issue (i.e. given _all supported (also extension provided form components, which is the one I support).

As part of #123 we obviously need to solve what you've sketched in https://github.com/bpmn-io/form-js/issues/237#issuecomment-1072018424. If you follow my line of thought (and we add a generic extension mechanism) then the form-js (editor or viewer) will sooner or later encounter elements in a schema that it just does not know. In this case we have only limited options:

* Viewer: Throw / indicate that the schema is not compatible; rationale: We don't want users ever to see and submit incomplete / partially rendered forms (a basic safety concern).

* Editor: Work like the viewer _or_ render a placeholder ("unknown" field) as you sketched in https://github.com/bpmn-io/form-js/issues/237#issuecomment-1072018424. I'd not attemt to even draw it, but simply render a placeholder "unknown field" or the like. UX rationale: As a user I am able to remove that form field; I cannot otherwise work with it though. This would allow users at least to fix their forms (if exported from an unknown form editor), which is a good property. However, let us not adopt it in the first place. Let us simply throw, if we encounter unknown elements, i.e. a FILE control in the stock form-js instance.

Hope this helps.

To sum up my proposal: Sketch a solution for #123 (be able to define [file upload](https://github.com/bpmn-io/form-js/pull/231) as an extension). Follow up with everything else.

Does that make sense?

**Comment by dfulgham** [comment]
@nikku Are you suggesting the ability for developers to add a form component via the new Form() or new FormEditor() properties, like passing an array of Component Objects into the constructor for example?

Or do you simply want a mechanism to make it easier to integrate new components into the code base? (ie. reorganize the code base and move components (editor and viewer combined) into their own lerna package?

**Comment by nikku** [comment]
> @nikku Are you suggesting the ability for developers to add a form component via the new Form() or new FormEditor() properties, like passing an array of Component Objects into the constructor for example?

I suggest exactly this.

**Comment by dfulgham** [comment]
@nikku see https://github.com/bpmn-io/form-js/issues/123#issuecomment-1073227631

**Comment by smbea** [comment]
Since the discussion in this issue seems relevant and ongoing, I will move it into backlog and keep the "help wanted" label.

**Comment by christian-konrad** [comment]
@nikku I updated the issue description as this is something beneficial for us if we now aim to add more components and properties that are not supported by previous Camunda execution engines

**Comment by nikku** [comment]
Makes sense as a limitation feature. We should not hide ship and hide experimental fields this way. Moving to backlog.
